docs: initial commit of Angular Package Format documentation (#43987)
angular/angular: aio/content/guide/angular-package-format.md

The versioning of the format is aligned with the versioning of Angular itself and we expect the format to evolve in a forward-compatible way to support the needs of the Angular component and tooling ecosystem.

## Why specify a package format?

In today's JavaScript landscape, developers will consume packages in many different ways, using many different toolchains (Webpack, rollup, esbuild, etc). These tools may understand and require different inputs - some tools may be able to process the latest ES language version, while others may benefit from directly consuming an older ES version.

The Angular distribution package supports all of the commonly used development tools and workflows, and adds emphasis on optimizations that result either in smaller application payload size or faster development iteration cycle (build time).

Developers can rely on Angular CLI and [ng-packagr](https://github.com/ng-packagr/ng-packagr) (a build tool Angular CLI uses) to produce packages in the Angular package format.

## File layout

The following example shows a simplified version of the `@angular/core` package's file layout, with an explanation for each file in the package.

> <code-example language="html">

Author: Converted to a table!

* Typings (`.d.ts` files)

`.d.ts` files are used by TypeScript when depending on a given package.

* `es2020` - ES2020 code flattened into a single `.mjs` file.
* `es2015` - ES2015 code flattened into a single `.mjs` file.
* `esm2020` - ES2020 code in unflattened `.mjs` files.

The ESM2020 format is included for the purpose of experimentation. Current bundlers are generally better at optimizing flattened files, but new tools are in development which may benefit from processing inputs in their original (unflattened) ES module format.

Tooling that is aware of these keys may preferentially select a desirable code format from `"exports"`. The remaining 2 keys control the default behavior of tooling:

* `"node"` selects flattened ES2015 code when the package is loaded in Node.

This format is used due to the requirements of `zone.js`, which does not support native `async`/`await` ES2017 syntax. Therefore, Node is instructed to use ES2015 code, where `async`/`await` structures have been downleveled into Promises.

> * `"default"` selects flattened ES2020 code for all other consumers.

Reviewer: In theory you could build your app without zone.js, in which case this is the more inclusive default. But I see how this could bite people that are building custom build pipelines - the good thing is that we now throw an error if you run your app in a way where async/await runs natively (without down-leveling).

> * `"default"` selects flattened ES2020 code for all other consumers.

Author: > then it is unlikely to have implemented the appropriate transformations to handle async-await properly.
I believe that we assume that whenever people are doing custom build/bundling, that they know what they're doing and will configure any necessary downleveling correctly.

It's only for Node, which usually runs code _unbundled_, that we have to explicitly provide code as ES2015.

This secondary `package.json` is required for legacy resolvers which do not support `"exports"`. 

## Entrypoints and Code Splitting

Packages in the Angular Package Format contain one primary entrypoint and zero or more secondary entrypoints (e.g. `@angular/common/http`). Entrypoints serve several functions.

1. They define the module specifiers from which users import code (for example, `@angular/core` and `@angular/core/testing`).

Users typically perceive these entrypoints as distinct groups of symbols, with different purposes or functionality.

Specific entrypoints might only be used for special purposes, such as testing. Such APIs can be separated out from the primary entrypoint to reduce the chance of them being used accidentally or incorrectly.

2. They define the granularity at which code can be lazily loaded.

Many modern build tools are only capable of "code splitting" (aka lazy loading) at the ES Module level. Since the Angular Package Format uses primarily a single "flat" ES Module per entrypoint, this means that most build tooling will not be able to split code in a single entrypoint into multiple output chunks.

> The general rule in for APF packages is to use entrypoints for the smallest sets of logically connected code possible. For example, the Angular Material package publishes each logical component or set of components as a separate entrypoint - one for Button, one for Tabs, etc. This allows each Material component to be lazily loaded separately, if desired.

Reviewer 2: Should this be "The general rule for APF packages"? (remove "in")

> Not all libraries require such granularity. Most libraries with a single logical purpose should be published as a single entrypoint. `@angular/core` for example uses a single entrypoint for the runtime, because the Angular runtime is generally used as a single entity.he exa

Reviewer 2: I think the characters after the period are unintentional.

More info: [webpack docs on side-effects](https://github.com/webpack/webpack/tree/master/examples/side-effects)
 
### ES2020 Language Level

ES2020 Language level is now the default language level that is consumed by Angular CLI and other tooling. The Angular CLI will downlevel the bundle to a language level that is supported by all targeted browsers at application build time.

### d.ts bundling / type definition flattening

> As of APF v8 we now prefer to run the .d.ts bundler tool from https://api-extractor.com/ so that the entire API appears in a single file.

Reviewer 3: As of APF v8 we prefer to run [API Extractor](https://api-extractor.com/), to bundle TypeScript definitions so that the entire API appears in a single file.

> As of APF v8 we now prefer to run the .d.ts bundler tool from https://api-extractor.com/ so that the entire API appears in a single file.

Reviewer 3: ```suggestion

As of APF v8 we prefer to run [API Extractor](https://api-extractor.com/), to bundle TypeScript definitions so that the entire API appears in a single file.
```

# Angular Package Format

> This document describes the Angular Package Format (APF).

Reviewer 4: I left a number of comments on a previous PR that turned the Google Doc into formal docs here: https://github.com/angular/angular/pull/37093. I think a fair number of these still apply.

Author: I actually incorporated most of those comments into this doc (at least, I think I hit most of them)

In prior APF versions each entry point would have a `src` directory next to the .d.ts entry point and this directory contained individual d.ts files matching the structure of the original source code. While this distribution format is still allowed and supported, it is highly discouraged because it confuses tools like IDEs that then offer incorrect autocompletion, and allows users to depend on deep-import paths which are typically not considered to be public API of a library or a package.

### Tslib

As of APF v10, we recommend adding tslib as a direct dependency of your primary entry-point. This is because the tslib version is tied to the TypeScript version used to compile your library.

## Examples

* [@angular/core package](https://unpkg.com/browse/@angular/core@13.0.0-rc.0/)
* [@angular/material package](https://unpkg.com/browse/@angular/material@13.0.0-rc.0/)

## Definition of Terms

The following terms are used throughout this document very intentionally. In this section we define all of them to provide additional clarity.

> <dl>

Reviewer: it seems that we don't have css for dl/dt/dd tags (or if we do have some then they are not good) because the output is poorly rendered and hard to read:

<img width="899" alt="Screen Shot 2021-11-01 at 3 55 54 PM" src="https://user-images.githubusercontent.com/216296/139752847-ada6ef0a-8cb9-4043-8fbc-9872e860b923.png">

Reviewer: I'm not quite sure what's the best way to fix this. Adding styles? turning this into a simple list? Moving most of these definitions to the glossary.md file?

Author: I added styles that make the section look much nicer - we can figure out if there's a better approach in a followup.

# Angular Package Format

This document describes the Angular Package Format (APF).

> APF is an Angular specific specification for layout of npm packages that is used by all first-party Angular packages and most third-party Angular libraries.

Reviewer: are we sure about "most"? wouldn't "many" be better?

> APF is an Angular specific specification for layout of npm packages that is used by all first-party Angular packages and most third-party Angular libraries.

Reviewer: sgtm

APF enables a package to work seamlessly under most common scenarios that use Angular.
Packages that use APF are compatible with the tooling offered by the Angular team as well as wider JavaScript ecosystem.
It is recommended that third-party library developers follow the same npm package format.

<div class="alert is-helpful">

APF is versioned along with the rest of Angular, and every major version improves the package format.
You can find the versions of the specification prior to v13 in this [google doc](https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/preview).

</div>

> This document describes the structure and format of the Angular framework packages currently available on npm. This format applies to packages distributing Angular components (like Angular Material) as well as the core framework packages published under the `@angular` namespace, such as `@angular/core` and `@angular/forms`.

Reviewer: yeah. I'd drop this paragraph.

> This document describes the structure and format of the Angular framework packages currently available on npm. This format applies to packages distributing Angular components (like Angular Material) as well as the core framework packages published under the `@angular` namespace, such as `@angular/core` and `@angular/forms`.

Author: Done, you're right, it was very redundant.

> This document describes the structure and format of the Angular framework packages currently available on npm. This format applies to packages distributing Angular components (like Angular Material) as well as the core framework packages published under the `@angular` namespace, such as `@angular/core` and `@angular/forms`.

Reviewer: great! thanks

## Why specify a package format?

In today's JavaScript landscape, developers will consume packages in many different ways, using many different toolchains (Webpack, rollup, esbuild, etc). These tools may understand and require different inputs - some tools may be able to process the latest ES language version, while others may benefit from directly consuming an older ES version.

The Angular distribution format supports all of the commonly used development tools and workflows, and adds emphasis on optimizations that result either in smaller application payload size or faster development iteration cycle (build time).

Developers can rely on Angular CLI and [ng-packagr](https://github.com/ng-packagr/ng-packagr) (a build tool Angular CLI uses) to produce packages in the Angular package format. See the [Creating Libraries](/guide/creating-libraries) guide for more details.

## File layout

The following example shows a simplified version of the `@angular/core` package's file layout, with an explanation for each file in the package. This table describes the file layout under `node_modules/@angular/core` annotated to describe the purpose of files and directories:

> <table>

Reviewer: this whole table doesn't render well on angular.io: https://pr43987-fc22f55.ngbuilds.io/guide/angular-package-format#file-layout

There is a lot of wasted space in cells and some cells are empty which makes the whole row look odd and redundant.

I wonder if it wouldn't be better to format this via either custom CSS or as monospaced code block (with ascii-art text layout).

Or at minimum move the whole thing to the bottom of the page? I dunno.. but we should do something about it — other ideas?

Reviewer 4: FWIW an ASCII layout probably isn't great for accessibility (vs. a list with nested ul/li)

Reviewer: right, we discussed this with Alex offline already. I agree.

<td>Code for all entrypoints in a flattened (FESM) ES2015 format, along with sourcemaps.</td>
</tr>
<tr>
<td><pre>fesm2015/core.mjs</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2015/core.mjs.map</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2015/testing.mjs</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2015/testing.mjs.map</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2020/</pre></td>
<td>Code for all entrypoints in flattened (FESM) ES2020 format, along with sourcemaps.</td>
</tr>
<tr>
<td><pre>fesm2020/core.mjs</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2020/core.mjs.map</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2020/testing.mjs</pre></td>
<td></td>
</tr>
<tr>
<td><pre>fesm2020/testing.mjs.map</pre></td>
<td></td>
</tr>
<tr>
<td><pre>testing/</pre></td>
<td>Directory representing the "testing" entrypoint.</td>
</tr>
<tr>
<td><pre>testing/package.json</pre></td>
<td>Defines the @angular/core/testing entrypoint with its typings for TypeScript, which does not support the top level "exports" based module resolution.</td>
</tr>
<tr>
<td><pre>testing/testing.d.ts</pre></td>
<td>Actual .d.ts for the @angular/core/testing entrypoint</td>
</tr>
</tbody>
</table>

## `package.json`

> Besides defining the package to NPM, the primary `package.json` performs several important functions:

Reviewer: ```suggestion
Besides defining the package to NPM, the primary `package.json` performs several important functions:
The primary `package.json` contains important package metadata, including the following:
```

> * It declares the package to be in EcmaScript Module (ESM) format, using `.mjs` files.

Reviewer: ```suggestion
* It declares the package to be in EcmaScript Module (ESM) format.
```

Reviewer: and link to the section below

> * It contains an `"exports"` field which defines the available source code formats of all entrypoints.

Reviewer: and link to the section below

* It contains keys which define the available source code formats of the primary `@angular/core` entrypoint, for tools which do not understand `"exports"`. These keys are considered deprecated, and will be removed as the support for `"exports"` rolls out across the ecosystem.

> * It declares the side-effectfulness of the package.

Reviewer: link to the section below.

### ESM declaration

The top-level `package.json` contains the key:

<code-example language="js">
{
  "type": "module"
}
</code-example>

> This informs resolvers that code within the package is using EcmaScript Modules as opposed to CommonJS modules, and uses the `.mjs` file extension.

Reviewer: is the ".mjs" comment right? I believe that with the `"type": "module"` set, the extension is non-loadbearing. We just happen to use .mjs because the bazel rules output that. But we can and might switch to `.js` extension in the future.

With the `"exports"` field set to `"module"`, the extension is an implementation detail not visible to package consumer.

Author: 👍 references to `.mjs` have been cleaned up.

"exports": {
  "./schematics/*": {
    "default": "./schematics/*.js"
  },
  "./package.json": {
    "default": "./package.json"
  },
  ".": {
    "types": "./core.d.ts",
    "esm2020": "./esm2020/core.mjs",
    "es2020": "./fesm2020/core.mjs",
    "es2015": "./fesm2015/core.mjs",
    "node": "./fesm2015/core.mjs",
    "default": "./fesm2020/core.mjs"
  },
  "./testing": {
    "types": "./testing/testing.d.ts",
    "esm2020": "./esm2020/testing/testing.mjs",
    "es2020": "./fesm2020/testing.mjs",
    "es2015": "./fesm2015/testing.mjs",
    "node": "./fesm2015/testing.mjs",
    "default": "./fesm2020/testing.mjs"
  }
}
</code-example>

Of primary interest are the `"."` and the `"./testing"` keys, which define the available code formats for the `@angular/core` primary entrypoint and the `@angular/core/testing` secondary entrypoint, respectively. For each entrypoint, the available formats are:

* Typings (`.d.ts` files)
`.d.ts` files are used by TypeScript when depending on a given package.
* `es2020` - ES2020 code flattened into a single `.mjs` file.
* `es2015` - ES2015 code flattened into a single `.mjs` file.
* `esm2020` - ES2020 code in unflattened `.mjs` files.

> The ESM2020 format is included for the purpose of experimentation. Current bundlers are generally better at optimizing flattened files, but new tools are in development which may benefit from processing inputs in their original (unflattened) ES module format.

Reviewer: link to the section below (and possibly shorten this explanation)

* `"node"` selects flattened ES2015 code when the package is loaded in Node.
This format is used due to the requirements of `zone.js`, which does not support native `async`/`await` ES2017 syntax. Therefore, Node is instructed to use ES2015 code, where `async`/`await` structures have been downleveled into Promises.
* `"default"` selects flattened ES2020 code for all other consumers.

### Legacy resolution keys

In addition to `"exports"`, the top-level `package.json` also defines legacy module resolution keys for resolvers that don't support `"exports"`. For `@angular/core` these are:

<code-example language="js">
{
  "fesm2020": "./fesm2020/core.mjs",
  "fesm2015": "./fesm2015/core.mjs",
  "esm2020": "./esm2020/core.mjs",
  "typings": "./core.d.ts",
  "module": "./fesm2015/core.mjs",
  "es2020": "./fesm2020/core.mjs",
}
</code-example>

As above, a module resolver can use these keys to load a specific code format. Note that instead of `"default"`, `"module"` selects the format both for Node as well as any tooling not configured to use a specific key. As with `"node"`, ES2015 code is selected due to the constraints of ZoneJS.

> ### Side effects

Reviewer: there is another whole section below on "sideEffects: false"- maybe shorten this one and move stuff down? or consider somehow combining the two please.

The last function of `package.json` is to declare whether the package has side-effects. By default, EcmaScript Modules are side-effectful: importing from a module ensures that any code at the top level of that module will execute. This is often undesirable, as most side-effectful code in typical modules is not truly side-effectful, but instead only affects specific symbols. If those symbols are not imported and used, it's often desirable to remove them in an optimization process known as tree-shaking, and the side-effectful code can prevent this.

> To work around this problem, packages can declare that they do not depend on side-effectful code at the top level of their modules, giving optimizers more freedom to tree-shake code from the package. This is done with another key in `package.json`:

Reviewer: Please link to https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free here or somewhere else in this section. thanks!

<code-example language="js">
{
  "sideEffects": false
}
</code-example>

Most Angular packages should not depend on top-level side effects, and thus should include this declaration.

### Secondary `package.json`s

In addition to the top-level `package.json`, secondary entrypoints have a corresponding directory with its own `package.json`. For example, in `@angular/core` the `@angular/core/testing` entrypoint is represented by `testing/package.json`.

> This secondary `package.json` is required for legacy resolvers which do not support `"exports"`. 

Reviewer: please add "And we expect these files to be removed from future versions of APF once more tools are updated to support the `"exports"` field.". thank you

## Entrypoints and Code Splitting

Packages in the Angular Package Format contain one primary entrypoint and zero or more secondary entrypoints (e.g. `@angular/common/http`). Entrypoints serve several functions.

1. They define the module specifiers from which users import code (for example, `@angular/core` and `@angular/core/testing`).

Users typically perceive these entrypoints as distinct groups of symbols, with different purposes or functionality.

> Specific entrypoints might only be used for special purposes, such as testing. Such APIs can be separated out from the primary entrypoint to reduce the chance of them being used accidentally or incorrectly.

Reviewer: this content should be offset, otherwise it renders poorly:

![Uploading Screen Shot 2021-11-01 at 3.51.25 PM.png…]()

2. They define the granularity at which code can be lazily loaded.

> Many modern build tools are only capable of "code splitting" (aka lazy loading) at the ES Module level. Since the Angular Package Format uses primarily a single "flat" ES Module per entrypoint, this means that most build tooling will not be able to split code in a single entrypoint into multiple output chunks.

Reviewer: same here
